pallets/gear-messenger/src/migrations.rs

```diff
@@ -0,0 +1,363 @@
+
+    impl From<ContextStore> for gear_core::message::ContextStore {
+        fn from(value: ContextStore) -> Self {
+            Self::new(
+                value.outgoing,
+                value.reply,
+                value.initialized,
+                value.reservation_nonce,
+                value.system_reservation,
+            )
+        }
+    }
+
+    pub struct DispatchesPrefix<T: Config>(PhantomData<T>);
+
+    impl<T: Config> StorageInstance for DispatchesPrefix<T> {
+        fn pallet_prefix() -> &'static str {
+            <<T as frame_system::Config>::PalletInfo as PalletInfo>::name::<Pallet<T>>()
+                .expect("No name found for the pallet in the runtime!")
+        }
+        const STORAGE_PREFIX: &'static str = "Dispatches";
+    }
+
+    impl<T: Config> CountedStorageMapInstance for DispatchesPrefix<T> {
+        type CounterPrefix = Self;
+    }
+
+    #[cfg(feature = "try-runtime")]
+    pub type Dispatches<T> = CountedStorageMap<
+        DispatchesPrefix<T>,
+        Identity,
+        MessageId,
+        LinkedNode<MessageId, StoredDispatch>,
+    >;
+
+    pub struct DispatchStashPrefix<T: Config>(PhantomData<T>);
+
+    impl<T: Config> StorageInstance for DispatchStashPrefix<T> {
+        fn pallet_prefix() -> &'static str {
+            <<T as frame_system::Config>::PalletInfo as PalletInfo>::name::<Pallet<T>>()
+                .expect("No name found for the pallet in the runtime!")
+        }
+        const STORAGE_PREFIX: &'static str = "DispatchStash";
+    }
+
+    #[cfg(feature = "try-runtime")]
+    #[allow(type_alias_bounds)]
+    pub type DispatchStash<T: Config> = StorageMap<
+        DispatchStashPrefix<T>,
+        Identity,
+        MessageId,
+        (StoredDispatch, Interval<BlockNumberFor<T>>),
+    >;
+
+    pub struct WaitlistPrefix<T: Config>(PhantomData<T>);
+
+    impl<T: Config> StorageInstance for WaitlistPrefix<T> {
+        fn pallet_prefix() -> &'static str {
+            <<T as frame_system::Config>::PalletInfo as PalletInfo>::name::<Pallet<T>>()
+                .expect("No name found for the pallet in the runtime!")
+        }
+        const STORAGE_PREFIX: &'static str = "Waitlist";
+    }
+
+    #[cfg(feature = "try-runtime")]
+    #[allow(type_alias_bounds)]
+    pub type Waitlist<T: Config> = StorageDoubleMap<
+        WaitlistPrefix<T>,
+        Identity,
+        ProgramId,
+        Identity,
+        MessageId,
+        (StoredDispatch, Interval<BlockNumberFor<T>>),
+    >;
+}
+
+#[cfg(test)]
+#[cfg(feature = "try-runtime")]
+mod test {
+    use crate::{
+        migrations::{v2, MigrateToV3},
+        mock::*,
+        DispatchStash, Dispatches, Waitlist,
+    };
+    use common::storage::{Interval, LinkedNode};
+    use frame_support::{pallet_prelude::StorageVersion, traits::OnRuntimeUpgrade};
+    use gear_core::{
+        ids::{MessageId, ProgramId},
+        message::{StoredDelayedDispatch, StoredMessage},
+    };
+
+    #[test]
+    fn migration_to_v3_works() {
```

Comment:
can't see @breathx suggestions on tests to be implemented

Reply:
Except they are? Compared to the old code, it generates three different StorageDispatches, some with Some and some with None, adds them to the old storage, migrates to the new one, then asserts that they are migrated correctly. I'm decently sure that is what he meant when he first mentioned my tests which just ran the migrations without any assertions. Or am I missing something?

Reply:
hmm ok, I will do that